Video refactor

penev92 opened the pull request:
The goal here was to change the output of video parsers (`IVideo.FrameData`) from `uint[,]` to a more standard `byte[]`.
Along the way I took the liberty to beautify and slightly refactor the interface and implementing classes.
Big and scary PR that should be reasonably separated by commits, so still best review by those.
~~Still has 2 issues to fix before it is done, both related to commit 3 - `Changed IVideo.CurrentFrameData [,] -> byte[]`.~~
The other 3 commits should be reviewable from the start, and all three are intended to not change anything other than code style, meaning there should be no behavioral changes.

P.S.: How would people feel about renaming `VqaReader` -> `VqaVideo` and `WsaReader` -> `WsaVideo`, since the interface they implement is called `IVideo` and they not only read, but also store (and to a limited sense "play") the video?

penev92 committed:
Did a beautification pass on IVideo and family

Removed property backing fields where applicable, introduced C#7 syntax for properties.
Renamed a bunch of interface properties and class private members with more descriptive names.
Did some inconsequential reordering.

penev92 committed:
Removed caching properties from video readers

Those seem redundant since the frame number is guaranteed to match the loaded data inside CurrentFrameData.

penev92 committed:
Changed IVideo.CurrentFrameData uint[,] -> byte[]

penev92 committed:
Reworked internal palettes in video reader classes

This removes the need to pack & unpack color bytes as uints for no gain.

penev92 committed:
Removed now-unused ITexture.SetData() overload

penev92 commented:
So commits 1-4 work nicely and let the video readers produce frame data as `byte[]` containing RGBA color data. Unfortunately the third commit (that does the actual `byte[]` change) breaks the `VideoPlayerWidget` and I haven't been able to fix that.
The 5th commit is a mix of cleaning up the now-unused `ITexture.SetData(uint[,])` overload and trying to fix the `VideoPlayerWidget`.
I guess I'm going to need help with the player widget.

penev92 committed:
Updated VideoPlayerWidget to play new IVideo data

Added optional padding to video frames because that's what VideoPlayerWidget expects.
Keeping the option to not use padding for other use-cases like converting frames to PNG.

penev92 committed:
Renamed IVideo implementations

To match the interface they are implementing.

penev92 commented:
This should now be completely reviewable.

Mailaender commented on OpenRA.Mods.Cnc/FileFormats/VqaVideo.cs:
```diff
@@ -160,12 +146,24 @@ public VqaReader(Stream stream)
 
 			CollectAudioData();
 
+			if (useFramePadding)
```
This is always true.

penev92 replied:
I am sure both this and the below comment were somehow tied to my experimental video-frame-to-png export (for playing videos in my experimental external asset browser, which started this entire PR), but can't currently find the example.

P.S.: I actually found an explanation in the commit's description:
```
Updated VideoPlayerWidget to play new IVideo data

Added optional padding to video frames because that's what VideoPlayerWidget expects.
Keeping the option to not use padding for other use-cases like converting frames to PNG.
```
So while all callers pass `true` at the moment, hopefully a follow-up PR will add a use-case that would pass `false`.

Mailaender commented on OpenRA.Mods.Cnc/FileFormats/VqaVideo.cs:
```diff
@@ -160,12 +146,24 @@ public VqaReader(Stream stream)
+			}
+			else
+			{
+				CurrentFrameData = new byte[Width * Height * 4];
```
This will probably crash when Width/Height are not power of 2.

penev92 replied:
As with the above comment, currently all callers are related to `VideoPlayerWidget`, which expects padding to be present, so they all pass `true` and go through the `if`. `VideoPlayerWidget` would indeed crash on this `else`, but creation of a PNG for example doesn't expect padding so that should be fine.

Mailaender reviewed (dismissed):
The videos still play and the interface changes as well as the minor refactoring make sense to me.

penev92 commented:
Rebased.

RoosterDragon commented on OpenRA.Game/Graphics/Video.cs:
```diff
@@ -13,13 +13,16 @@ namespace OpenRA.Video
+		/// Current frame color data in 32-bit BGRA.
+		/// </summary>
+		byte[] CurrentFrameData { get; }
+		int CurrentFrameNumber { get; }
```
Overthinking/nitpicking: `CurrentFrameIndex`? `Index` strongly suggests 0-indexing to me, whereas `Number` feels more 1-indexey.

RoosterDragon commented on OpenRA.Game/Graphics/VideoLoader.cs:
```diff
@@ -15,15 +15,15 @@ namespace OpenRA.Video
 	public interface IVideoLoader
 	{
-		bool TryParseVideo(Stream s, out IVideo video);
+		bool TryParseVideo(Stream s, bool useFramePadding, out IVideo video);
```
Thought: Is it worth documenting the effects `useFramePadding`? I for one was surprised it's the max of both dimensions rounded up to nearest power of two, as opposed to each dimension individually rounded up to next power of 2.

e.g. Currently 100x10 gives a padding of 128x128, whereas individually would be 128x16.

penev92 replied:
How/where would you recommend I document this though? It would probably have to be an XML comment on `IVideoLoader`, but it doesn't feel right there. *(Actually the padding is a special case for the `VideoPlayerWidget` sooo ugh...)*

RoosterDragon commented on OpenRA.Mods.Cnc/FileFormats/VqaVideo.cs:
```diff
@@ -16,31 +16,32 @@
+		public int CurrentFrameNumber { get; private set; }
+
+		public bool HasAudio { get; set; }
+		public byte[] AudioData { get; private set; } // audio for this frame: 22050Hz 16bit mono pcm, uncompressed.
```
Feels reasonable to promote this to an xml comment.

penev92 replied:
I am not convinced actually - such a comment should generally go on `IVideo`, but this particular comment is specific to the `VqaVideo` implementation of `IVideo`. And we also never expose `VqaVideo` publicly, so if the comment is not on the interface it would still not be visible anywhere.

RoosterDragon commented on OpenRA.Mods.Common/Widgets/VideoPlayerWidget.cs:
```diff
@@ -93,27 +94,29 @@ public override void Draw()
 
 				if (skippedFrames > 1)
-					Log.Write("perf", "VqaPlayer : {0} skipped {1} frames at position {2}", cachedVideo, skippedFrames, video.CurrentFrame);
+					Log.Write("perf", "VideoPlayer: {0} skipped {1} frames at position {2}", cachedVideo, skippedFrames, video.CurrentFrameNumber);
```
`nameof(VideoPlayerWidget)`?

RoosterDragon commented on OpenRA.Platforms.Default/ThreadedGraphicsContext.cs:
```diff
@@ -557,7 +557,6 @@ class ThreadedTexture : ITextureInternal
 		readonly Func<object> getSize;
 		readonly Action<object> setEmpty;
 		readonly Func<byte[]> getData;
-		readonly Func<object, object> setData1;
```
You probably kept it simple to reduce the diff size, but I think renumbering the fields would be nice.

RoosterDragon reviewed (dismissed):
LGTM, comments for consideration but not required.

penev92 commented:
Sadly these comments all make sense to me.

Mailaender commented:
Needs a rebase.

penev92 committed:
Addressed review comments

 - Renamed `IVideo.CurrentFrameNumber` to `CurrentFrameIndex`
 - Improved logged error message in VideoPlayerWidget
 - Renumbered fields in ThreadedGraphicsContext

penev92 commented:
Rebased and fixed some of the requests.
I see no way we can make the two comment requests work, so unless anyone has more suggestions I'll skip them.
